segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ZKSegmentContainerMonitor.java

```diff
@@ -196,51 +202,63 @@ private void checkAssignment() {
         this.pendingTasks.add(containerId);
-        return this.registry
-                .startContainer(containerId, INIT_TIMEOUT_PER_CONTAINER)
-                .whenComplete((handle, ex) -> {
-                    try {
-                        if (ex == null) {
-                            if (this.handles.putIfAbsent(handle.getContainerId(), handle) != null) {
-                                log.warn("Starting container {} succeeded but handle is already registered.",
-                                        handle.getContainerId());
+        try {
+            return this.registry
+                    .startContainer(containerId, INIT_TIMEOUT_PER_CONTAINER)
+                    .whenComplete((handle, ex) -> {
+                        try {
+                            if (ex == null) {
+                                if (this.handles.putIfAbsent(handle.getContainerId(), handle) != null) {
+                                    log.warn("Starting container {} succeeded but handle is already registered.",
+                                            handle.getContainerId());
+                                } else {
+                                    handle.setContainerStoppedListener(this::unregisterHandle);
+                                    log.info("Container {} has been registered.", handle.getContainerId());
+                                }
                             } else {
-                                handle.setContainerStoppedListener(this::unregisterHandle);
-                                log.info("Container {} has been registered.", handle.getContainerId());
+                                log.warn("Starting container {} failed: {}", containerId, ex);
                             }
-                        } else {
-                            log.warn("Starting container {} failed: {}", containerId, ex);
+                        } finally {
+                            // The pending task has to be removed in the end to avoid inconsistencies since containerhandle
+                            // should be available immediately after the task is complete.
+                            // Also need to ensure this is always called, hence doing this in a finally block.
+                            this.pendingTasks.remove(containerId);
                         }
-                    } finally {
-                        // The pending task has to be removed in the end to avoid inconsistencies since containerhandle
-                        // should be available immediately after the task is complete.
-                        // Also need to ensure this is always called, hence doing this in a finally block.
-                        this.pendingTasks.remove(containerId);
-                    }
-                });
+                    });
+        } catch (Throwable e) {
+            // The pending task has to be removed on all failures to enable retries.
+            this.pendingTasks.remove(containerId);
```

Comment:
The pending tasks state is used to ensure no 2 operations for the same container happen in parallel. We check this state before attempting to start/stop a container. So if container 1 has a previously pending task (starting/stopping), we postpone the start/stop operation to the next iteration of this monitor task.